app/clients/letter/dvla.py

```diff
@@ -125,19 +144,20 @@ def jwt_token(self):
 
         return self._jwt_token
 
-    def _handle_common_dvla_errors(self, e: requests.HTTPError):
-        if e.response.status_code == 429:
-            raise DvlaThrottlingException() from e
-        elif e.response.status_code >= 500:
-            raise DvlaRetryableException() from e
-        else:
-            raise DvlaNonRetryableException() from e
-
     def authenticate(self):
         """
         Fetch a JWT from the DVLA API that can be used in other DVLA API requests
         """
-        try:
+
+        def _handle_401(e: requests.HTTPError):
+            if e.response.status_code == 401:
+                # likely the old password has already expired
+                current_app.logger.exception("Failed to generate a DVLA jwt token")
+
+                self.dvla_password.clear()
+                raise DvlaRetryableException(e.response.json()[0]["detail"]) from e
+
+        with _handle_common_dvla_errors(custom_httperror_exc_handler=_handle_401):
```

Comment:
interested to hear how people feel about this. is it obvious and understandable what this does and how we'd update/write another block in the future?

Reply:
Feels reasonable to me 👍  I think if I'd been doing it I'd have ended up adding a `post` method to the class (which takes an optional error handler) and centralising the logic there rather than using a context manager. Would remove a level of indentation, and mean that when adding future methods/calls to the client, we don't need to remember to add the context manager as all post requests would get the error handling natively.

But I don't think it is problematic enough to actually change what you've done.

Important part is that what you've got does what we need, so nice 👍